toolkit/tools/precacher/precacher.go

```diff
@@ -0,0 +1,426 @@
+			logger.Log.Infof("Pre-caching: %3d%% ( downloaded: %4d, skipped: %4d, unavailable: %4d, failed: %4d )", int(progressPercent), downloaded, skipped, unavailable, failed)
+			lastProgressUpdate = progressPercent
+		}
+	}
+	return
+}
+
+// precachePackage will attempt to download the specified package. It will return a downloadResult struct via the results.
+// This function runs with best effort, so it will return all errors via the results channel. rather than returning an error.
+// The results may be one of:
+//   - downloadResultTypeSuccess: The package was downloaded successfully
+//   - downloadResultTypeFailure: The package failed to download (ie error occured)
+//   - downloadResultTypeSkipped: The package was not downloaded because it already exists
+//   - donwloadResultTypeUnavailable: The package was not downloaded because it was not found in any of the repos
+//
+// The caller is expected to have added to the provided wait group, while this function is
+// responsible for removing itself from the wait group. As much processing as possible is done before acquiring the
+// network operations semaphore to minimize the time spent holding it.
+func precachePackage(pkg *repocloner.RepoPackage, packagesAvailableFromRepos map[string]string, outDir string, wg *sync.WaitGroup, results chan<- downloadResult, netOpsSemaphore chan struct{}) {
+	const (
+		downloadRetryAttempts = 2
+		downloadRetryDuration = time.Second
+	)
+
+	defer wg.Done()
+
+	// File names are of the form "<name>-<version>.<distro>.<arch>.rpm"
+	pkgName := fmt.Sprintf("%s-%s.%s.%s", pkg.Name, pkg.Version, pkg.Distribution, pkg.Architecture)
+	fileName := fmt.Sprintf("%s.rpm", pkgName)
+	fullFilePath := path.Join(outDir, fileName)
+	result := downloadResult{
+		pkgName: fileName,
+	}
+
+	// Bail out early if the file already exists
+	if exists, _ := file.PathExists(fullFilePath); exists {
+		result.resultType = downloadResultTypeSkipped
+		results <- result
+		return
+	}
+
+	// Get the url for the package, or bail out if it is not available
+	url, ok := packagesAvailableFromRepos[pkgName]
+	if !ok {
+		result.resultType = donwloadResultTypeUnavailable
+		results <- result
+		return
+	}
+
+	// Limit the number of concurrent network operations to avoid overloading the network. All work past this point
+	// is either network related, or trivial, so we can safely hold the semaphore until we are done.
+	netOpsSemaphore <- struct{}{}
+	defer func() {
+		<-netOpsSemaphore
+	}()
+
+	logger.Log.Debugf("Pre-caching '%s' from '%s'", fileName, url)
+	err := retry.Run(func() error {
+		err := network.DownloadFile(url, fullFilePath, nil, nil)
+		if err != nil {
+			logger.Log.Warnf("Attempt to download (%s) failed. Error: %s", url, err)
+		}
+		return err
+	}, downloadRetryAttempts, downloadRetryDuration)
+
+	if err != nil {
+		result.resultType = downloadResultTypeFailure
+		results <- result
+		return
+	}
+
+	result.resultType = downloadResultTypeSuccess
+	results <- result
+}
```

Comment:
Idea to make sure we:
- never forget to send a result,
- educe the number of times we have to do it,
- and make erroring-out require the least amount of changes.

A different default would be also OK, but it's more common to error-out in multiple places, while other statuses are probably more likely to be set only one time.

```suggestion
	const (
		downloadRetryAttempts = 2
		downloadRetryDuration = time.Second
	)

	// File names are of the form "<name>-<version>.<distro>.<arch>.rpm"
	pkgName := fmt.Sprintf("%s-%s.%s.%s", pkg.Name, pkg.Version, pkg.Distribution, pkg.Architecture)
	fileName := fmt.Sprintf("%s.rpm", pkgName)
	fullFilePath := path.Join(outDir, fileName)
	result := downloadResult{
		pkgName: fileName,
		resultType: downloadResultTypeFailure,
	}

	defer defer func() {
		results <- result
		wg.Done()
	}()

	// Bail out early if the file already exists
	if exists, _ := file.PathExists(fullFilePath); exists {
		result.resultType = downloadResultTypeSkipped
		return
	}

	// Get the url for the package, or bail out if it is not available
	url, ok := packagesAvailableFromRepos[pkgName]
	if !ok {
		result.resultType = donwloadResultTypeUnavailable
		return
	}

	// Limit the number of concurrent network operations to avoid overloading the network. All work past this point
	// is either network related, or trivial, so we can safely hold the semaphore until we are done.
	netOpsSemaphore <- struct{}{}
	defer func() {
		<-netOpsSemaphore
	}()

	logger.Log.Debugf("Pre-caching '%s' from '%s'", fileName, url)
	err := retry.Run(func() error {
		err := network.DownloadFile(url, fullFilePath, nil, nil)
		if err != nil {
			logger.Log.Warnf("Attempt to download (%s) failed. Error: %s", url, err)
		}
		return err
	}, downloadRetryAttempts, downloadRetryDuration)
	if err != nil {
		return
	}

	result.resultType = downloadResultTypeSuccess
}
```